<!DOCTYPE html>


  <html class="light page-post">


<head>
  <meta charset="utf-8">
  
  <title>基因注释 | tigerose&#39;s site</title>

  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">

  
    <meta name="keywords" content="基因注释," />
  

  <meta name="description" content="基因注释记录下自己对RNA-seq基因注释的学习,并对Drop-seq软件包中的注释模块进行代码研读 什么是基因注释一句话概况注释:找到与reads有overlap的基因片段,并进行标记 这里reads指bam文件中的每一行数据,即测序下机文件fastq与参考基因组进行比对之后生成的数据,其中记录了每条read在参考基因组中的位置,有起始位置和终止位置,表示一段区间 基因注释文件记录了每个基因片段">
<meta property="og:type" content="article">
<meta property="og:title" content="基因注释">
<meta property="og:url" content="http://yoursite.com/2020/04/30/gene-annotation/index.html">
<meta property="og:site_name" content="tigerose&#39;s site">
<meta property="og:description" content="基因注释记录下自己对RNA-seq基因注释的学习,并对Drop-seq软件包中的注释模块进行代码研读 什么是基因注释一句话概况注释:找到与reads有overlap的基因片段,并进行标记 这里reads指bam文件中的每一行数据,即测序下机文件fastq与参考基因组进行比对之后生成的数据,其中记录了每条read在参考基因组中的位置,有起始位置和终止位置,表示一段区间 基因注释文件记录了每个基因片段">
<meta property="article:published_time" content="2020-04-30T12:49:05.000Z">
<meta property="article:modified_time" content="2020-04-30T12:59:22.580Z">
<meta property="article:author" content="tigerose">
<meta property="article:tag" content="基因注释">
<meta name="twitter:card" content="summary">

  

  
    <link rel="icon" href="/images/favicon.ico">
  

  <link href="/css/styles.css?v=c114cbeddx" rel="stylesheet">


  
    
<link rel="stylesheet" href="/css/personal-style.css">

  

  

  
  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "//hm.baidu.com/hm.js?d4b892153a53af797abac43f46aac4dc";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>


  
  <script type="text/javascript">
	(function(){
	    var bp = document.createElement('script');
	    var curProtocol = window.location.protocol.split(':')[0];
	    if (curProtocol === 'https') {
	        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
	    }
	    else {
	        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
	    }
	    var s = document.getElementsByTagName("script")[0];
	    s.parentNode.insertBefore(bp, s);
	})();
  </script>



  
    <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <link rel="stylesheet" href="//cdn.bootcss.com/font-awesome/4.3.0/css/font-awesome.min.css">
  

<meta name="generator" content="Hexo 4.2.0"><link rel="alternate" href="/atom.xml" title="tigerose's site" type="application/atom+xml">
</head>

<body>


  
    <span id="toolbox-mobile" class="toolbox-mobile">盒子</span>
  

  <div class="post-header CENTER">
   
  <div class="toolbox">
    <a class="toolbox-entry" href="/">
      <span class="toolbox-entry-text">盒子</span>
      <i class="icon-angle-down"></i>
      <i class="icon-home"></i>
    </a>
    <ul class="list-toolbox">
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/archives/"
            rel="noopener noreferrer"
            target="_self"
            >
            博客
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/category/"
            rel="noopener noreferrer"
            target="_self"
            >
            分类
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/tag/"
            rel="noopener noreferrer"
            target="_self"
            >
            标签
          </a>
        </li>
      
    </ul>
  </div>


</div>


  <div id="toc" class="toc-article">
    <strong class="toc-title">文章目录</strong>
    <ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#基因注释"><span class="toc-text">基因注释</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#什么是基因注释"><span class="toc-text">什么是基因注释</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#基因注释文件"><span class="toc-text">基因注释文件</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#注释流程分析"><span class="toc-text">注释流程分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Drop-seq代码研读"><span class="toc-text">Drop-seq代码研读</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#主流程"><span class="toc-text">主流程</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#加载gtf并构建interval-tree"><span class="toc-text">加载gtf并构建interval tree</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#注释逻辑"><span class="toc-text">注释逻辑</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#输出统计信息"><span class="toc-text">输出统计信息</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#参考"><span class="toc-text">参考</span></a></li></ol></li></ol>
  </div>



<div class="content content-post CENTER">
   <article id="post-gene-annotation" class="article article-type-post" itemprop="blogPost">
  <header class="article-header">
    <h1 class="post-title">基因注释</h1>

    <div class="article-meta">
      <span>
        <i class="icon-calendar"></i>
        <span>2020.04.30</span>
      </span>

      
        <span class="article-author">
          <i class="icon-user"></i>
          <span>tigerose</span>
        </span>
      

      


      

      
      <i class="fa fa-eye"></i> 
        <span id="busuanzi_container_page_pv">
           &nbsp阅读 <span id="busuanzi_value_page_pv">
           <i class="fa fa-spinner fa-spin"></i></span>
        </span>
      
      
    </div>
  </header>

  <div class="article-content">
    
      <h1 id="基因注释"><a href="#基因注释" class="headerlink" title="基因注释"></a>基因注释</h1><p>记录下自己对RNA-seq基因注释的学习,并对Drop-seq软件包中的注释模块进行代码研读</p>
<h2 id="什么是基因注释"><a href="#什么是基因注释" class="headerlink" title="什么是基因注释"></a>什么是基因注释</h2><p>一句话概况注释:<em>找到与reads有overlap的基因片段,并进行标记</em></p>
<p>这里reads指bam文件中的每一行数据,即测序下机文件fastq与参考基因组进行比对之后生成的数据,其中记录了每条read在参考基因组中的位置,有起始位置和终止位置,表示一段区间</p>
<p>基因注释文件记录了每个基因片段在参考基因组上的位置,也是一段区间,因此与bam文件结合,通过find overlapping我们可以查找到每条read属于哪个基因片段,将其标记在bam格式的tags中,这对后续的生信分析是有帮助的</p>
<h2 id="基因注释文件"><a href="#基因注释文件" class="headerlink" title="基因注释文件"></a>基因注释文件</h2><p>GTF/GFF格式是基因注释的常用格式<br>GTF是Gene Transfer Format的缩写,其文件由九列数据组成,以tab分割,示例如下:</p>
<table>
<thead>
<tr>
<th>seq_id</th>
<th>source</th>
<th>type</th>
<th>start</th>
<th>end</th>
<th>score</th>
<th>strand</th>
<th>phase</th>
<th>attributes</th>
</tr>
</thead>
<tbody><tr>
<td>chr1</td>
<td>HAVANA</td>
<td>exon</td>
<td>11869</td>
<td>12227</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>gene_id “ENSG00000223972.5”; transcript_id “ENST00000456328.2”;…</td>
</tr>
<tr>
<td>染色体编号</td>
<td>注释来源</td>
<td>类型</td>
<td>在参考序列的起始位置</td>
<td>终止位置</td>
<td>得分,说明注释信息的可能性</td>
<td>位于参考序列的正负链</td>
<td>仅对类型为CDS有效,表示起始编码的位置</td>
<td>包含众多属性的列表</td>
</tr>
</tbody></table>
<p>虽然数据有九列之多,但并不是所有都会用到,常用的有:</p>
<ol>
<li>seq_id. 要查找overlap,首先得是同一条染色体</li>
<li>type. 有多种类型,如gene/transcript/exon/CDS/UTR等,它们之间有层级关系,一般gtf文件中多行数据对应一条基因的完整信息,以type为gene的行为起始;每条gene可以表示为树状结构,gene为根节点,第二层为transcript,第三层为诸如exon CDS等</li>
<li>start/end. 根据起始终止位置可以建立interval,这是find overlapping的基础</li>
<li>strand. 正负链可以作为过滤条件,假如一条read与多个基因有overlap,可以根据方向是否相同过滤掉部分基因</li>
<li>attributes. 一些列键值对属性,常用的信息包括名称,id之类</li>
</ol>
<h2 id="注释流程分析"><a href="#注释流程分析" class="headerlink" title="注释流程分析"></a>注释流程分析</h2><p>流程可分为三步:</p>
<ol>
<li><strong>读入gtf文件</strong>. 从磁盘将gtf文件加载进内存,并提取需要的信息,毕竟gtf有许多信息是我们不需要的</li>
<li><strong>建立区间树</strong>. 即interval tree,使用区间树是为了高效查询,为了达到最佳性能,一般使用基于红黑树的区间树实现,因为红黑树是平衡树,查找时间复杂度O(lgN),不会出现退化成链表的最坏情况</li>
<li><strong>查找区间并注释</strong>. 遍历bam文件中每条read,根据其在参考序列中的位置构建interval,与前面建立的interval tree进行overlap的查找,找到之后,进行一些逻辑计算,并更新read的tags,输出到bam</li>
</ol>
<h2 id="Drop-seq代码研读"><a href="#Drop-seq代码研读" class="headerlink" title="Drop-seq代码研读"></a>Drop-seq代码研读</h2><p>Drop-seq是使用java开发的程序包,其中的<em>TagReadWithGeneExonFunction</em> 模块实现了添加注释的功能</p>
<h3 id="主流程"><a href="#主流程" class="headerlink" title="主流程"></a>主流程</h3><p>主流程核心代码为:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 加载基因注释文件,并构建区间树</span></span><br><span class="line"><span class="keyword">final</span> OverlapDetector&lt;Gene&gt; geneOverlapDetector = GeneAnnotationReader.loadAnnotationsFile(ANNOTATIONS_FILE, bamDict);</span><br><span class="line"><span class="comment">// 打开输出bam文件</span></span><br><span class="line">SAMFileWriter writer= <span class="keyword">new</span> SAMFileWriterFactory().makeSAMOrBAMWriter(header, <span class="keyword">true</span>, OUTPUT);</span><br><span class="line"><span class="comment">// 遍历输入bam中的每条read</span></span><br><span class="line"><span class="keyword">for</span> (SAMRecord r: inputSam) &#123;</span><br><span class="line">    <span class="comment">// 对完成比对的read,进行find overlapping操作并添加注释</span></span><br><span class="line">    <span class="keyword">if</span> (!r.getReadUnmappedFlag())</span><br><span class="line">        r=setAnnotations(r, geneOverlapDetector);</span><br><span class="line">    <span class="comment">// 输出注释后的read</span></span><br><span class="line">    writer.addAlignment(r);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>其结果是根据overlap的genes信息,添加三个Tag,示例:</p>
<blockquote>
<p>GE:Z:WASH7P    XF:Z:CODING        GS:Z:-</p>
</blockquote>
<ul>
<li>GE为gene name  </li>
<li>XF为locus function  </li>
<li>GS为正负链</li>
</ul>
<h3 id="加载gtf并构建interval-tree"><a href="#加载gtf并构建interval-tree" class="headerlink" title="加载gtf并构建interval tree"></a>加载gtf并构建interval tree</h3><p>核心代码:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 用来解析gtf每行数据,提取需要的字段</span></span><br><span class="line"><span class="keyword">final</span> FilteringGTFParser parser = <span class="keyword">new</span> FilteringGTFParser(gtfFlatFile);</span><br><span class="line"><span class="comment">// gene name相同的gtf行,代表它们是一个gene的数据,使用它们构建GeneFromGTF数据类型,</span></span><br><span class="line"><span class="comment">// 此类型继承自Interval类型</span></span><br><span class="line"><span class="keyword">final</span> GeneFromGTFBuilder geneBuilder = <span class="keyword">new</span> GeneFromGTFBuilder(parser);</span><br><span class="line"><span class="comment">// 初始化interval tree</span></span><br><span class="line"><span class="keyword">final</span> OverlapDetector&lt;GeneFromGTF&gt; overlapDetector = <span class="keyword">new</span> OverlapDetector&lt;&gt;(<span class="number">0</span>, <span class="number">0</span>);</span><br><span class="line"><span class="keyword">while</span> (geneBuilder.hasNext())&#123;</span><br><span class="line">        <span class="comment">// 将每条gene添加到interval tree,其内部按照chromosome进行分类</span></span><br><span class="line">        GeneFromGTF gene = geneBuilder.next();</span><br><span class="line">        overlapDetector.addLhs(gene, gene);</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>

<p>代码实现使用了迭代器,减少内存的消耗</p>
<p>将gtf每行数据以gene_name为key,放入<code>map&lt;gene_name, List&lt;GTFRecord&gt;&gt;</code>中,这样就将每条gene的数据分类好了<br>geneBuilder 是个<code>iter&lt;List&lt;GTFRecord&gt;&gt;</code>,迭代时,对每个gene将其数据<code>List&lt;GTFRecord&gt;</code> 按gene_version分类成<code>map&lt;gene_version, List&lt;GTFRecord&gt;&gt;</code>,对key也就是所有的gene_version进行排序,取最大的gene_version对应的<code>List&lt;GTFRecord&gt;</code>来构建GeneFromGTF(对部分gtf文件它们的gene_version是null,则对gene_version分类和没做一样)</p>
<p>重点是构建GeneFromGTF类实例:<br><code>List&lt;GTFRecord&gt;</code>转GeneFromGTF调用接口<code>makeGeneFromMultiVersionGTFRecords()</code></p>
<ol>
<li>使用list中第一条GTFRecord的信息初始化GeneFromGTF(因为第一条的类型永远是gene),只有start end属性是取得list中所有数据的最小start,最大end  </li>
<li>进行一致性检查. 检查list中所有数据,如正反链必须都一致,chr一致等,否则抛出异常</li>
<li>将所有的非gene数据进行统计处理,更新GeneFromGTF成员变量<code>Map&lt;String, TranscriptFromGTF&gt; transcripts</code>, TranscriptFromGTF类型包括transcript, CDS, coding等相关信息</li>
</ol>
<p>最后调用<em>htsjdk</em>库的<code>OverlapDetector.addLhs()</code>将GeneFromGTF作为节点加入线段树中</p>
<h3 id="注释逻辑"><a href="#注释逻辑" class="headerlink" title="注释逻辑"></a>注释逻辑</h3><p>核心代码:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">public</span> SAMRecord <span class="title">setAnnotations</span> <span class="params">(<span class="keyword">final</span> SAMRecord r, <span class="keyword">final</span> OverlapDetector&lt;Gene&gt; geneOverlapDetector)</span> </span>&#123;</span><br><span class="line">        Map&lt;Gene, LocusFunction&gt; map = AnnotationUtils.getInstance().getLocusFunctionForReadByGene(r, geneOverlapDetector);</span><br><span class="line">        Set&lt;Gene&gt; exonsForRead = AnnotationUtils.getInstance().getConsistentExons (r, map.keySet(), ALLOW_MULTI_GENE_READS);</span><br><span class="line"></span><br><span class="line">        List&lt;Gene&gt; genes = <span class="keyword">new</span> ArrayList&lt;&gt;();</span><br><span class="line"></span><br><span class="line">        <span class="keyword">for</span> (Gene g: exonsForRead) &#123;</span><br><span class="line">            LocusFunction f = map.get(g);</span><br><span class="line">            <span class="keyword">if</span> (f==LocusFunction.CODING || f==LocusFunction.UTR)</span><br><span class="line">                genes.add(g);</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">        List&lt;LocusFunction&gt; allPassingFunctions = <span class="keyword">new</span> ArrayList&lt;&gt;();</span><br><span class="line">        <span class="keyword">if</span> (USE_STRAND_INFO) &#123;</span><br><span class="line">            <span class="comment">// constrain gene exons to read strand.</span></span><br><span class="line">            genes = getGenesConsistentWithReadStrand(genes, r);</span><br><span class="line">            <span class="comment">// only retain functional map entries that are on the correct strand.</span></span><br><span class="line">            <span class="keyword">for</span> (Gene g: map.keySet()) &#123;</span><br><span class="line">                <span class="keyword">boolean</span> strandCheck=readAnnotationMatchStrand(g, r);</span><br><span class="line">                <span class="keyword">if</span> (strandCheck) allPassingFunctions.add(map.get(g));</span><br><span class="line">            &#125;</span><br><span class="line">        &#125; <span class="keyword">else</span></span><br><span class="line">            allPassingFunctions=<span class="keyword">new</span> ArrayList&lt;&gt;(map.values());</span><br><span class="line"></span><br><span class="line">        <span class="comment">// if strand tag is used, only add locus function values for passing genes.</span></span><br><span class="line">        <span class="keyword">for</span> (Gene g: genes)</span><br><span class="line">            allPassingFunctions.add(map.get(g));</span><br><span class="line"></span><br><span class="line">        LocusFunction f = AnnotationUtils.getInstance().getLocusFunction(allPassingFunctions, <span class="keyword">false</span>);</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> (genes.size()&gt;<span class="number">1</span> &amp;&amp; <span class="keyword">this</span>.ALLOW_MULTI_GENE_READS==<span class="keyword">false</span>)</span><br><span class="line">            log.error(<span class="string">"There should only be 1 gene assigned to a read for DGE purposes."</span>);</span><br><span class="line"></span><br><span class="line">        String finalGeneName = getCompoundGeneName(genes);</span><br><span class="line">        String finalGeneStrand = getCompoundStrand(genes);</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> (f!=<span class="keyword">null</span>)</span><br><span class="line">            r.setAttribute(<span class="keyword">this</span>.FUNCTION_TAG, f.toString());</span><br><span class="line">        <span class="keyword">if</span> (finalGeneName!=<span class="keyword">null</span> &amp;&amp; finalGeneStrand!=<span class="keyword">null</span>) &#123;</span><br><span class="line">            r.setAttribute(<span class="keyword">this</span>.TAG, finalGeneName);</span><br><span class="line">            r.setAttribute(<span class="keyword">this</span>.STRAND_TAG, finalGeneStrand);</span><br><span class="line">        &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">            r.setAttribute(<span class="keyword">this</span>.TAG, <span class="keyword">null</span>);</span><br><span class="line">            r.setAttribute(<span class="keyword">this</span>.STRAND_TAG, <span class="keyword">null</span>);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">return</span> (r);</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>

<p>概况一下注释逻辑:对read构建interval,查找overlap的所有基因,计算三个tag: locusFunction, geneName, geneStrand,并更新read  </p>
<p>对每个gene计算locusFunction的核心代码:</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">private</span> LocusFunction <span class="title">getLocusFunctionForRead</span> <span class="params">(<span class="keyword">final</span> SAMRecord rec, <span class="keyword">final</span> Gene g)</span> </span>&#123;</span><br><span class="line">    List&lt;AlignmentBlock&gt; alignmentBlocks = rec.getAlignmentBlocks();</span><br><span class="line"></span><br><span class="line">    LocusFunction [] blockSummaryFunction = <span class="keyword">new</span> LocusFunction[alignmentBlocks.size()];</span><br><span class="line">    Set&lt;Gene&gt; temp = <span class="keyword">new</span> HashSet&lt;&gt;();</span><br><span class="line">    temp.add(g);</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> (<span class="keyword">int</span> i=<span class="number">0</span>; i&lt;alignmentBlocks.size(); i++) &#123;</span><br><span class="line">        AlignmentBlock alignmentBlock =alignmentBlocks.get(i);</span><br><span class="line"></span><br><span class="line">        LocusFunction [] blockFunctions=getLocusFunctionsByBlock(alignmentBlock, temp);</span><br><span class="line">        LocusFunction blockFunction = getLocusFunction(blockFunctions, <span class="keyword">false</span>);</span><br><span class="line">        blockSummaryFunction[i]=blockFunction;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    LocusFunction readFunction = getLocusFunction(blockSummaryFunction, <span class="keyword">false</span>);</span><br><span class="line">    <span class="keyword">return</span> readFunction;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>计算过程:</p>
<ol>
<li>对read每个位点都计算一个LocusFunction(计算过程具体为是否与exons有overlap)</li>
<li>将一条alignmentBlocks中所有位点的LocusFunctions合并为一个LocusFunction</li>
<li>将所有alignmentBlocks的LocusFunctions合并为一个LocusFunction</li>
</ol>
<p>每次合并都是取最大的LocusFunction,其是一个枚举变量,由小到大为:<br><code>INTERGENIC, INTRONIC, UTR, CODING, RIBOSOMAL</code></p>
<h3 id="输出统计信息"><a href="#输出统计信息" class="headerlink" title="输出统计信息"></a>输出统计信息</h3><table>
<thead>
<tr>
<th>统计项</th>
<th>解释</th>
</tr>
</thead>
<tbody><tr>
<td>AMBIGUOUS_READS_REJECTED</td>
<td>有多条同方向的overlaped gene</td>
</tr>
<tr>
<td>READ_AMBIGUOUS_GENE_FIXED</td>
<td>overlaped gene中同向的只有一条,但是还有反向的</td>
</tr>
<tr>
<td>READS_RIGHT_STRAND</td>
<td>只有一条overlaped gene,并且与read同向</td>
</tr>
<tr>
<td>READS_WRONG_STRAND</td>
<td>overlaped gene没有同向,只有反向</td>
</tr>
<tr>
<td>TOTAL_READS</td>
<td>处理的总reads数</td>
</tr>
</tbody></table>
<h2 id="参考"><a href="#参考" class="headerlink" title="参考"></a>参考</h2><ul>
<li><a href="https://blog.csdn.net/sinat_38163598/article/details/72851239" target="_blank" rel="noopener">GTF基因注释文件详解</a></li>
<li><a href="https://www.jianshu.com/p/0800a07cfa37" target="_blank" rel="noopener">液滴微流控获取单细胞及Drop-seq_tools分析流程</a></li>
<li><a href="https://github.com/broadinstitute/Drop-seq" target="_blank" rel="noopener">Drop-seq_github</a></li>
</ul>

    
  </div>

</article>


   

   
  <div class="box-prev-next clearfix">
    <a class="show pull-left" href="/2020/04/26/intro-seqan/">
        <i class="icon icon-angle-left"></i>
    </a>
    <a class="hide pull-right" href="/">
        <i class="icon icon-angle-right"></i>
    </a>
  </div>




</div>


  <a id="backTop" class="back-top">
    <i class="icon-angle-up"></i>
  </a>




  <div class="modal" id="modal">
  <span id="cover" class="cover hide"></span>
  <div id="modal-dialog" class="modal-dialog hide-dialog">
    <div class="modal-header">
      <span id="close" class="btn-close">关闭</span>
    </div>
    <hr>
    <div class="modal-body">
      <ul class="list-toolbox">
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/archives/"
              rel="noopener noreferrer"
              target="_self"
              >
              博客
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/category/"
              rel="noopener noreferrer"
              target="_self"
              >
              分类
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/tag/"
              rel="noopener noreferrer"
              target="_self"
              >
              标签
            </a>
          </li>
        
      </ul>

    </div>
  </div>
</div>



  
      <div class="fexo-comments comments-post">
    

    

    
    

    

    
    

    

  </div>

  

  <script type="text/javascript">
  function loadScript(url, callback) {
    var script = document.createElement('script')
    script.type = 'text/javascript';

    if (script.readyState) { //IE
      script.onreadystatechange = function() {
        if (script.readyState == 'loaded' ||
          script.readyState == 'complete') {
          script.onreadystatechange = null;
          callback();
        }
      };
    } else { //Others
      script.onload = function() {
        callback();
      };
    }

    script.src = url;
    document.getElementsByTagName('head')[0].appendChild(script);
  }

  window.onload = function() {
    loadScript('/js/bundle.js?235683', function() {
      // load success
    });
  }
</script>

</body>
</html>
